templates/fine-tune-llm_v2/end-to-end-examples/fine-tune-preference/README.ipynb

```diff
@@ -0,0 +1,1071 @@
+    "\n",
+    "![preprocessing](./assets/preprocessing.png?1)\n",
+    "\n",
+    "💡 INSIGHT: \n",
+    "Our synthetic preference data collection looks pretty involved at first glance. The key ideas in plain English are as follows:\n",
+    "- We use a combination of Q&A scoring + word length to indicate like/dislike (our preference function) given a pair of summaries.\n",
+    "- Our ultimate goal is to generate (chosen, rejected) pairs to train our reference model and evaluate it based on this criteria.\n",
+    "- We use another LLM (judge model) to generate said questions for each article. This model is also used in our scoring system. (To see how many questions can be answered from a summary)\n",
+    "- To generate training data, we first sample candidate summaries from the reference model for each article. We then obtain scores for each summary from the judge. Using the scores, we select pairs of summaries and mark our like/dislike to form (chosen, rejected) pairs for the actual training. \n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Part (a): Generate Multiple Choice Questions from Articles\n",
+    "\n",
+    "First, we will generate the multiple choice questions and answers for each article using `Llama-3.1-8B-Instruct` (or `70B` if A100/H100s are available). Leveraging vLLM and Ray, we can very easily scale this generation process across multiple GPUs.\n",
+    "\n",
+    ">  **_NOTE:_**  We provide two sets of configs: One with an 8B parameter model as the judge, and another with the 70B model. Using the 8B model is recommended, since we make use of highly available A10Gs. For good performance, and to replicate the results in our blog, you should use the 70B judge model which uses A100s (but these are harder to obtain on-demand)\n",
+    "\n",
+    "The following command will run the [src/scripts/generate_questions.py](./src/scripts/generate_questions.py) script, which generates the questions and answers and saves them in `.parquet` files."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "\n",
+    "💡 INSIGHT:  \n",
+    "We are running this script as an anyscale job. The resources required by each step are requested at runtime and provisioned by Anyscale's autoscaler based on availability and quotas. You are free to change the [qa_generation](./configs/qa_generation) config in any way. The important parameters regarding resources are `accelerator_type`, `num_gpus_per_instance`, and `concurrency`. This script will generate 5 multiple choice question and answer pairs per article for 21k examples. According to the [llama_8b](./configs/qa_generation/llama_8b.yaml) config we are requesting 3 replicas of 4xA10G machines processing a batch-size of 128 examples each which saturates the GPUs all the way through.\n",
+    "\n",
+    "This step will take ~40 min for 8B running on 12 A10s (~ $10) and ~75 mins for 70B running on 8 A100s (~ $28). \n",
```

Comment:
the format is messed up.

Reply:
hello yes will ping when ready for review :)